### PR TITLE
docs(subscriber): console-subscriber API docs

### PR DESCRIPTION
**Pull request by pnkfelix**
Part of #194: document public API of the console-subscriber crate.

### CODE REVIEW

File: console-subscriber/src/builder.rs [hunk 1]
```diff
@@ -83,7 +83,12 @@ impl Builder {
     /// A shorter duration will allow clients to update more frequently, but may
     /// result in the program spending more time preparing task data updates.
     ///
-    /// By default, this is [`TasksLayer::DEFAULT_PUBLISH_INTERVAL`].
+    /// By default, this is [`TasksLayer::DEFAULT_PUBLISH_INTERVAL`], though
```
Comment (hawkw) [inline console-subscriber/src/builder.rs]:
```suggestion
    /// By default, this is [`ConsoleLayer::DEFAULT_PUBLISH_INTERVAL`], though
```

File: console-subscriber/src/builder.rs [hunk 2]
```diff
@@ -98,15 +103,26 @@ impl Builder {
     /// will reduce memory usage, but less historical data from completed tasks
     /// will be retained.
     ///
-    /// By default, this is [`TasksLayer::DEFAULT_RETENTION`].
+    /// By default, this is [`TasksLayer::DEFAULT_RETENTION`], though methods
```
Comment (hawkw) [inline console-subscriber/src/builder.rs]:
```suggestion
    /// By default, this is [`ConsoleLayer::DEFAULT_RETENTION`], though methods
```

File: console-subscriber/src/lib.rs [hunk 3]
```diff
@@ -38,6 +38,11 @@ pub use builder::{init, spawn};
 use crate::aggregator::Id;
 use crate::visitors::{PollOpVisitor, StateUpdateVisitor};
 
+/// A [`TasksLayer`] is a [`tracing_subscriber::Layer`] that catches the tracing
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]:
```suggestion
/// A [`ConsoleLayer`] is a [`tracing_subscriber::Layer`] that catches the tracing
```

File: console-subscriber/src/builder.rs [hunk 4]
```diff
@@ -83,7 +83,12 @@ impl Builder {
     /// A shorter duration will allow clients to update more frequently, but may
     /// result in the program spending more time preparing task data updates.
     ///
-    /// By default, this is [`ConsoleLayer::DEFAULT_PUBLISH_INTERVAL`].
+    /// By default, this is [`ConsoleLayer::DEFAULT_PUBLISH_INTERVAL`], though
+    /// methods like [`init`][`crate::init`] and [`spawn`][`crate::spawn`] will
+    /// take the value from the `TOKIO_CONSOLE_PUBLISH_INTERVAL` [environment
+    /// variable] before falling back on that default.
```
Comment (hawkw) [inline console-subscriber/src/builder.rs]: how about breaking this into multiple sentences?
```suggestion
    /// By default, this is [`ConsoleLayer::DEFAULT_PUBLISH_INTERVAL`].
    /// Methods like [`init`][`crate::init`] and [`spawn`][`crate::spawn`] will
    /// take the value from the `TOKIO_CONSOLE_PUBLISH_INTERVAL` [environment
    /// variable] before falling back on that default.
```

File: console-subscriber/src/builder.rs [hunk 5]
```diff
@@ -98,15 +103,26 @@ impl Builder {
     /// will reduce memory usage, but less historical data from completed tasks
     /// will be retained.
     ///
-    /// By default, this is [`ConsoleLayer::DEFAULT_RETENTION`].
+    /// By default, this is [`ConsoleLayer::DEFAULT_RETENTION`], though methods
+    /// like [`init`][`crate::init`] and [`spawn`][`crate::spawn`] will take the
```
Comment (hawkw) [inline console-subscriber/src/builder.rs]:
```suggestion
    /// By default, this is [`ConsoleLayer::DEFAULT_RETENTION`]. Methods
    /// like [`init`][`crate::init`] and [`spawn`][`crate::spawn`] will take the
```

File: console-subscriber/src/builder.rs [hunk 6]
```diff
@@ -98,15 +103,26 @@ impl Builder {
...
+    /// [environment variable]: `Builder::with_default_env`
     pub fn retention(self, retention: Duration) -> Self {
         Self { retention, ..self }
     }
 
     /// Sets the socket address on which to serve the RPC server.
     ///
     /// By default, the server is bound on the IP address [`Server::DEFAULT_IP`]
-    /// on port [`Server::DEFAULT_PORT`].
+    /// on port [`Server::DEFAULT_PORT`], though methods like
```
Comment (hawkw) [inline console-subscriber/src/builder.rs]:
```suggestion
    /// on port [`Server::DEFAULT_PORT`]. Methods like
```

File: console-subscriber/src/builder.rs [hunk 7]
```diff
@@ -115,6 +131,13 @@ impl Builder {
     }
 
     /// Sets the path to record the events to the file system.
+    ///
+    /// By default, this is initially `None`, though methods like
+    /// [`init`][`crate::init`] and [`spawn`][`crate::spawn`] will take the
+    /// value from the `TOKIO_CONSOLE_RECORD_PATH` [environment variable] before
+    /// falling back on that default.
```
Comment (hawkw) [inline console-subscriber/src/builder.rs]:
```suggestion
    /// By default, this is initially `None`. Methods like
    /// [`init`][`crate::init`] and [`spawn`][`crate::spawn`] will take the
    /// value from the `TOKIO_CONSOLE_RECORD_PATH` [environment variable] before
    /// falling back on that default.
```

File: console-subscriber/src/lib.rs [hunk 8]
```diff
@@ -38,6 +38,11 @@ pub use builder::{init, spawn};
 use crate::aggregator::Id;
 use crate::visitors::{PollOpVisitor, StateUpdateVisitor};
 
+/// A [`ConsoleLayer`] is a [`tracing_subscriber::Layer`] that catches the tracing
+/// events of the runtime (e.g. call-sites), and translating them into more
+/// abstract events (e.g. task spawns or waker operations) that are meaningful
+/// to modeling the async runtime, and then transmitting those abstact events
+/// to an associated [`Server`] for aggregation.
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]: this is one _very_ long sentence, can we split it up? here's a suggestion:

```suggestion
/// A [`ConsoleLayer`] is a [`tracing_subscriber::Layer`] that records [`tracing`]
/// spans and events emitted by the async runtime. 
///
/// Runtimes emit [`tracing`] spans and events that represent specific operations
/// that occur in asynchronous Rust programs, such as spawning tasks and waker
/// operations. The `ConsoleLayer` collects and aggregates these events, and the
/// resulting diagnostic data is exported to clients by the corresponding gRPC 
/// [`Server`] instance.
///
/// [`tracing`]: https://docs.rs/tracing
```

File: console-subscriber/src/lib.rs [hunk 9]
```diff
@@ -94,6 +99,12 @@ pub struct ConsoleLayer {
     no_dispatch: Dispatch,
 }
 
+/// A [`Server`] is in charge of aggregating the abstract runtime events into a
+/// model of the runtime's history (such as a list of the currently active
+/// tasks, or statistics summarizing polling times). A [`Server`] also
+/// interprets commands from a client application, such a request to focus in on
+/// a specific task, and translates that into a stream of details specific to
+/// that task.
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]: technically the server doesn't perform the aggregation, this is done in a separate task that both the layer and the server talk to. i would prefer to focus on what this does in the API; how about something like this?
```suggestion
/// A gRPC [`Server`] that implements the [`tokio-console` wire format][wire].
///
/// Client applications, such as the [`tokio-console CLI][cli] connect to the gRPC
/// server, and stream data about the runtime's history (such as a list of the 
/// currently active tasks, or statistics summarizing polling times). A [`Server`] also
/// interprets commands from a client application, such a request to focus in on
/// a specific task, and translates that into a stream of details specific to
/// that task.
///
/// [wire]: https://docs.rs/console-api
/// [cli]: https://crates.io/crates/tokio-console
```

File: console-subscriber/src/lib.rs [hunk 10]
```diff
@@ -272,11 +293,40 @@ impl ConsoleLayer {
 }
 
 impl ConsoleLayer {
+    /// Default maximum capacity for channel of events from [`ConsoleLayer`] to [`Server`].
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]:
```suggestion
    /// Default maximum capacity for the channel of events sent from a
    /// [`ConsoleLayer`] to a [`Server`].
```

File: console-subscriber/src/lib.rs [hunk 11]
```diff
@@ -272,11 +293,40 @@ impl ConsoleLayer {
...
+    ///
+    /// When this capacity is exhausted, additional events will be dropped.
```
Comment (hawkw) [inline console-subscriber/src/lib.rs:309]:
```suggestion
    /// When this capacity is exhausted, additional events will be dropped.
    /// Decreasing this value will reduce memory usage, but may result in
    /// events being dropped more frequently.
```

File: console-subscriber/src/lib.rs [hunk 12]
```diff
@@ -272,11 +293,40 @@ impl ConsoleLayer {
...
+    ///
+    /// See also [`Builder::event_buffer_capacity`].
     pub const DEFAULT_EVENT_BUFFER_CAPACITY: usize = 1024 * 10;
+    /// Default maximum capacity for channel of events from [`Server`] to each subscribed client.
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]:
```suggestion
    /// Default maximum capacity for th echannel of events sent from a
    /// [`Server`] to each subscribed client.
```

File: console-subscriber/src/lib.rs [hunk 13]
```diff
@@ -272,11 +293,40 @@ impl ConsoleLayer {
...
+    ///
+    /// When this capacity is exhausted, the client may be disconnected.
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]:
```suggestion
    /// When this capacity is exhausted, the client is assumed to be inactive,
    /// and may be disconnected.
```

File: console-subscriber/src/lib.rs [hunk 14]
```diff
@@ -272,11 +293,40 @@ impl ConsoleLayer {
...
+    ///
+    /// See also [`Builder::client_buffer_capacity`].
     pub const DEFAULT_CLIENT_BUFFER_CAPACITY: usize = 1024 * 4;
+
+    /// Default frequency for publication of events to clients.
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]:
```suggestion
    /// Default frequency for publishing events to clients.
```

File: console-subscriber/src/lib.rs [hunk 15]
```diff
@@ -641,13 +691,40 @@ impl fmt::Debug for ConsoleLayer {
...
     pub const DEFAULT_PORT: u16 = 6669;
 
+    /// Initiates service via the default gRPC transport service.
     pub async fn serve(self) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
         self.serve_with(tonic::transport::Server::default()).await
     }
 
+    /// Initiates service via the given gRPC transport service `builder`.
+    ///
+    /// Both event aggregation and client command interpretation services are initiated here.
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]:
```suggestion
    /// Starts the gRPC service with the given [`tonic`] gRPC transport server
    /// `builder`.
    ///
    /// The `builder` parameter may be used to configure gRPC-specific settings
    /// prior to starting the server.
    ///
    /// This spawns both the server task and the event aggregation worker
    /// task on the current async runtime.
    ///
    /// [`tonic`]: https://docs.rs/tonic/
```

File: console-subscriber/src/lib.rs [hunk 16]
```diff
@@ -641,13 +691,40 @@ impl fmt::Debug for ConsoleLayer {
...
+    /// [`spawn`][`crate::spawn`] will parse the socket address from the
+    /// `TOKIO_CONSOLE_BIND` [environment variable] before falling back on
+    /// constructing a socket address from this default.
+    ///
+    /// See also [`Builder::server_addr`].
+    ///
+    /// [environment variable]: `Builder::with_default_env`
     pub const DEFAULT_PORT: u16 = 6669;
 
+    /// Initiates service via the default gRPC transport service.
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]: how about:
```suggestion
    /// Starts the gRPC service with the default gRPC settings.
    ///
    /// To configure gRPC server settings before starting the server, use 
    /// [`serve_with`] instead. This method is equivalent to calling [`serve_with`]
    /// and providing the default gRPC server settings:
    ///
    /// ```rust
    /// # async fn docs() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
    /// # let (_, server) = console_subscriber::ConsoleLayer::new();
    /// server.serve_with(tonic::transport::Server::default()).await
    /// # }
    /// ```
```

**Review by hawkw — CHANGES_REQUESTED** [review]

**Review by hawkw — APPROVED** [review]

File: console-subscriber/src/lib.rs [hunk 17]
```diff
@@ -641,13 +706,58 @@ impl fmt::Debug for ConsoleLayer {
...
+    /// To configure gRPC server settings before starting the server, use
+    /// [`serve_with`] instead. This method is equivalent to calling [`serve_with`]
+    /// and providing the default gRPC server settings:
+    ///
+    /// ```rust
+    /// # async fn docs() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    /// # let (_, server) = console_subscriber::ConsoleLayer::new();
+    /// server.serve_with(tonic::transport::Server::default()).await
+    /// # }
+    /// ```
```
Comment (hawkw) [inline console-subscriber/src/lib.rs:746]: i guess rustdoc can't resolve this link? https://app.netlify.com/sites/tokio-console/deploys/61ba771eb07ff90007653aa9
```suggestion
    /// ```
    /// [`serve_with`]: Builder::serve_with
```

Comment (pnkfelix) [reply console-subscriber/src/lib.rs:746]: Do need to put quotes around `Builder::serve_with`, like:
```
/// [`serve_with`]: `Builder::serve_with`
```
?

Comment (carllerche) [reply console-subscriber/src/lib.rs:746]: @pnkfelix not for doc links. That should be enough.

File: console-subscriber/src/lib.rs [hunk 18]
```diff
@@ -641,13 +706,59 @@ impl fmt::Debug for ConsoleLayer {
...
+    /// [`serve_with`]: Builder::serve_with
```
Comment (hawkw) [inline console-subscriber/src/lib.rs]: BLAH wrong type
```suggestion
    /// [`serve_with`]: Server::serve_with
```

**Comment by carllerche** [comment]
greeeeeeen :green_apple:

**Comment by hawkw** [comment]
:raised_hands:

**Pull request merged**